### ISSUE TITLE
ArrayInputFactory fails in production mode

### ISSUE THREAD

**Issue by giacmir** [magento/magento2#7604]
I wrote a shell command that imports products from a custom format. At the end of the import process I need to call two other commands to launch a reindex and to empty cache.

To do so I get an Symfony\Component\Console\Input\ArrayInputFactory from my constructor and then I create the ArrayInput object in this way:

```
$parameters = [
  'command' => 'indexer:reindex',
  'index' => ['cataloginventory_stock'],
 ];

$indexerArgs = $this->arrayInputFactory->create(['parameters' => $parameters]);
```

This used to work fine in 2.0 but in 2.1 it works only in developer mode. In production mode the command dies with the following error:

```
PHP Fatal error:  Uncaught TypeError: Argument 1 passed to Symfony\Component\Console\Input\ArrayInput::__construct() must be of the type array, object given, called in /var/www/nuna_stage/web/releases/20161128162520/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php on line 93 and defined in /var/www/nuna_stage/web/releases/20161128162520/vendor/symfony/console/Symfony/Component/Console/Input/ArrayInput.php:37
Stack trace:
#0 /var/www/nuna_stage/web/releases/20161128162520/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(93): Symfony\Component\Console\Input\ArrayInput->__construct(Object(Magento\Framework\ObjectManager\ObjectManager))
#1 /var/www/nuna_stage/web/releases/20161128162520/vendor/magento/framework/ObjectManager/Factory/Compiled.php(88): Magento\Framework\ObjectManager\Factory\AbstractFactory->createObject('Symfony\\Compone...', Array)
#2 /var/www/nuna_stage/web/releases/20161128162520/vendor/magento/framework/ObjectManager/ObjectManager.php(57): Magento\Framework\ObjectManager\Fact in /var/www/nuna_stage/web/releases/20161128162520/vendor/symfony/console/Symfony/Component/Console/Input/ArrayInput.php on line 37
```

### Preconditions
1. Magento 2.1.2
2. Php 7

### Steps to reproduce

1. Create a new command
2. Use ArrayInputFactory as described above
3. launch the command

### Expected result

1. The command runs without errors both in developer and in production mode

### Actual result

1. In developer mode the command runs, in production it fails.

**Comment by giacmir** [comment]
Note that this same exception is mentioned in #2604 and that issue was closed with a "put your site in developer mode before running the command". That could be valid for installing sample data but not in my case obviously.

**Comment by shiftedreality** [comment]
Hi @giacmir 

We've created internal ticket MAGETWO-62676 to investigate and fix this issue.
As a temporary solution you can create an `ArrayInput` object via `new` keyword:
```php
$commandInput = new ArrayInput($arguments);
```

**Comment by giacmir** [comment]
Thanks @shiftedreality 

In fact using `new` is exactly what I've done just after opening this issue. :wink:

**Comment by phoenix128** [comment]
Same error here with custom class. Seems related to `\Magento\Framework\ObjectManager\Factory\Compiled` class not passing correct arguments

**Comment by magento-engcom-team** [comment]
@giacmir, thank you for your report.
The issue is already fixed in develop branch, 2.2.0
